open-ce/build_tree.py

```diff
@@ -194,6 +216,50 @@ def _add_build_command_dependencies(build_commands, start_index=0):
                 deps += filter(lambda x: x != start_index + index, packages[dep])
         build_command.build_command_dependencies = deps
 
+def _remove_duplicate_build_commands(variant_recipes, build_commands):
+    """
+    Identify duplicate build_commands and remove them.
+
+    Update the `build_command_dependencies` for each build_command.
+    """
+    if build_commands:
+        start_index = len(build_commands)
+
+        #TODO: remove debug prints
+        # Create dictionary listing the index of entries that are present both in build_commands
+        # and in variant_recipes
+        duplicates = {}
+        for index, entry in enumerate(variant_recipes):
+            if entry in build_commands:
+                print("EXISTING entry at INDEX %s in buildcommands  matches entry at index %s in varrecipes"
+                       % (build_commands.index(entry), index + start_index ))
+                duplicates[index + start_index] = build_commands.index(entry)
+        print(duplicates)
+
+        print("---------REPLACING index variant recipes-----------")
+        # Replace duplicate indices in `build_command_dependencies` with corresponding original indices.
+        if duplicates:
+            for build_command in variant_recipes:
+                print("command deps %s" % (build_command.build_command_dependencies))
+                updated_command_deps = []
+                for i in build_command.build_command_dependencies:
+                    updated_command_deps.append(duplicates.get(i,i))
+                build_command.build_command_dependencies = updated_command_deps
+                print("updated deps %s" % (build_command.build_command_dependencies))
+            print("--------------------------------------")
+
+        # Remove duplicate build_commands and update indices in `build_command_dependencies`
+        for i, key in enumerate(duplicates):
```

Comment:
This looks like a pretty expensive operation (possibly n^3). I think we will need a new way to organize the build commands to avoid this. One possibility is to change the build_commands value to be a dictionary instead of a list. This way, when you remove items from the container it won't disrupt the indices.